src/backend/distributed/commands/table.c

```diff
@@ -593,59 +609,185 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
 
 
 /*
- * ErrorIfAlterTableDefinesFKeyFromPostgresToCitusLocalTable errors out if
- * given ALTER TABLE statement defines foreign key from a postgres local table
- * to a citus local table.
+ * AlterTableDefinesFKeyBetweenPostgresAndNonDistTable returns true if given
+ * alter table command defines foreign key between a postgres table and a
+ * reference or citus local table.
  */
-static void
-ErrorIfAlterTableDefinesFKeyFromPostgresToCitusLocalTable(
-	AlterTableStmt *alterTableStatement)
+static bool
+AlterTableDefinesFKeyBetweenPostgresAndNonDistTable(AlterTableStmt *alterTableStatement)
 {
-	List *commandList = alterTableStatement->cmds;
+	bool rightHasPostgresLocalTable = false;
+	bool rightHasNonDistTable = false;
 
-	LOCKMODE lockmode = AlterTableGetLockLevel(commandList);
-	Oid leftRelationId = AlterTableLookupRelation(alterTableStatement, lockmode);
+	Oid rightRelationId = InvalidOid;
+	List *rightRelationIdList =
+		GetAlterTableAddFKeyRightRelationIdList(alterTableStatement);
+	foreach_oid(rightRelationId, rightRelationIdList)
+	{
+		if (!IsCitusTable(rightRelationId))
+		{
+			rightHasPostgresLocalTable = true;
+		}
+		else if (IsCitusTableType(rightRelationId, CITUS_TABLE_WITH_NO_DIST_KEY))
+		{
+			rightHasNonDistTable = true;
+		}
+	}
 
-	if (IsCitusTable(leftRelationId))
+	LOCKMODE lockmode = AlterTableGetLockLevel(alterTableStatement->cmds);
+	Oid leftRelationId = AlterTableLookupRelation(alterTableStatement, lockmode);
+	bool leftIsPostgresTable = !IsCitusTable(leftRelationId);
+	if (leftIsPostgresTable)
 	{
-		/* left relation is not a postgres local table, */
-		return;
+		return rightHasNonDistTable;
 	}
 
-	List *alterTableFKeyConstraints =
-		GetAlterTableStmtFKeyConstraintList(alterTableStatement);
-	Constraint *constraint = NULL;
-	foreach_ptr(constraint, alterTableFKeyConstraints)
+	bool leftIsNonDistTable = IsCitusTableType(leftRelationId,
+											   CITUS_TABLE_WITH_NO_DIST_KEY);
+	if (leftIsNonDistTable)
 	{
-		Oid rightRelationId = RangeVarGetRelid(constraint->pktable, lockmode,
-											   alterTableStatement->missing_ok);
-		if (IsCitusTableType(rightRelationId, CITUS_LOCAL_TABLE))
+		return rightHasPostgresLocalTable;
+	}
+
+	return false;
+}
+
+
+/*
+ * ConvertPostgresLocalTablesToCitusLocalTables converts each postgres table
+ * involved in foreign keys to be defined by given alter table command and the
+ * other tables connected to them via a foreign key graph to citus local tables.
+ */
+static void
+ConvertPostgresLocalTablesToCitusLocalTables(AlterTableStmt *alterTableStatement)
```

Comment:
What about `alter table add column references`, should we handle that as well?

Reply:
yes, we actually handle add column commands via `GetAlterTableCommandFKeyConstraintList` function

Reply:
See my other comment, I think I was confused because of that